### DISCUSSION TITLE
ingame_python contrib, failed tests, successful game interaction

### DISCUSSION THREAD

**Discussion by jamalainm** [evennia/evennia#2332] [Evennia Questions] [answered]
Hi all,

I've followed the directions for the installation of 'ingame_python' as laid out in the README.

That doesn't seem to be problematic at all, and upon switching over the typeclasses to "Event*" the in-game experience seems fine.

However, when I run my old tests, the "say" command now fails the test. The message that the test expects is empty, but in-game it works just fine. Here's the Fail message:

```
Traceback (most recent call last):
  File ".../NoviōraTempora/mygame/commands/tests.py", line 284, in test_say
    self.call(general.CmdSay(), "Testing", 'You say, "Testing"')
  File ".../NoviōraTempora/evennia/evennia/commands/default/tests.py", line 165, in call
    raise AssertionError(retval)
AssertionError: 
==============================Wanted message==================================
You say, "Testing"
==============================Returned message================================

==============================================================================
```

Has anyone else experienced this sort of thing with the ingame_python contrib?

Thanks for your time!

**Comment by davewiththenicehat** [comment]
I had an issue recently because I was using MULTISESSION_MODE =2.
The AssertionError was identical, but I don't think I had an issue with CmdSay.

In short some of the unit tests log in, but don't puppet a Character.
To run the built in tests I imported the evennia tests, and ones with this issue I added: `self.account.puppet_object(self.session, self.char1)` to the top of the test.

Perfectly possible you are seeing something completely different. Though no message on a built in test, there is a chance it is this issue. Only if you use multisession 2, maybe 3. I don't know enough about the modes to say for certain.

**Reply by jamalainm** [reply]
Hrmm. Including self.account.puppet_object(self.session, self.char1) didn't lead to different test results for me. I did have multisession mode at 2, but even doing away with multisession mode leaves me with the empty return message.

Thanks for the suggestion, though!

**Comment by Griatch** [comment, marked as answer]
The `ingame_python` contrib overwrites `at_say` on your Character (in `EventCharacter`) and whereas the normal `at_say` calls various message receivers through `.msg` (which is what the default CommandTest will mock in order for you to get something to compare), the `EventCharacter.at_say` will set off a series of callbacks which may or may not eventually lead to a `msg` call, but in a very different location the mock knows nothing about. Hence the mock completely misses whatever is returned.

The issue here is likely that `self.char1` on the CommandTest base is using your Character class (now set to `EventCharacter`). It should use DefaultCharacter in order to always work with the default code. On the Evennia side we could hardcode this of course, but it would instead make it harder for people to use this base class for their own unit tests, so ... it's a balancing act. 

Maybe we should make one `DefaultCommandTest` class only for default tests and another one using whatever Character class is currently active.

**Reply by jamalainm** [reply]
That's a really good idea! Thanks for also explaining why my test was getting the results it was!
